### PR TITLE
feat: #305 forest client api backend integration

### PR DESCRIPTION
**Pull request by ianliuwk1019**
Details for #305 PR:
* Add a class responsible for interfacing (make api calls to) Forest Client API.
* Use API Token from environment variable for setting up the request header.
* Add new FAM Forest Client - search endpoint by query parameter (client_number).
  - guarded with authorized user login.
  - for now restrict to exact 8-digits length (limitation from Forest Client API)
* Parse API return result for FAM search endpoint needs: "client_name", "forest_client_number", "status: code/description".
* Create custom crud business validator to validate creat user/role assignment before saving:
  - validate if the forest client exists.
  - validate if the forest client is active (ACT).
* Generate API Client for frontend use.
* Add new tests.
* API Token Setup for all environments. (Keys are set at repo first as suggested by Conrad)
* Adjust CI workflow so it can read GH environment secret token for Unit Testing.

### CODE REVIEW

File: server/backend/api/app/crud/crud_user_role.py [hunk 1]
```diff
@@ -215,3 +229,42 @@ def find_or_create_forest_client_child_role(
...
+            fc_api = ForestClient()
+
+            # Locally stored (if any) for later use to prevent api calls again.
+            # Exact client number search - should only contain 1 result.
+            self.fc = fc_api.find_by_client_number(forest_client_number)
+            LOGGER.debug(f"Forest Client(s) retrieved: {self.fc}")
+
+    def forest_client_number_exists(self) -> bool:
+        # Exact client number search - should only contain 1 result.
+        return len(self.fc) == 1
```
Comment (basilv) [inline server/backend/api/app/crud/crud_user_role.py:259]: So multiple results will cause this to fail, I'm thinking I'm okay with this.

**Review by basilv — CHANGES_REQUESTED** [review]

File: server/backend/api/app/crud/crud_forest_client.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
+import json
```
[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/backend/api/app/crud/crud_forest_client.py:1]: Good point, thanks Basil!
I guess I have 'crud' already baked in to hobby but in this case, I guess I can just moved extra code into router itself.

[the remark this answers is not in the document]
Comment (gormless87) [reply server/backend/api/app/crud/crud_forest_client.py:1]: Leftover from some Guy named Fran?

[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/backend/api/app/crud/crud_forest_client.py]: Updated.
Hahah no Pete, it's leftover from the Empire master 'Fran's' apprentice. From now on, I should look at your Python coding style guide instead :P

File: client-code-gen/gen/model/fam-forest-client-status.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+     * @type {FamForestClientStatusType}
+     * @memberof FamForestClientStatus
+     */
+    'status_code': FamForestClientStatusType;
+    /**
+     * 
+     * @type {string}
+     * @memberof FamForestClientStatus
+     */
+    'description': string;
```
Comment (MCatherine1994) [inline client-code-gen/gen/model/fam-forest-client-status.ts:37]: just wondering shall we give the description a real description? or it doesn't matter?

**Review by basilv — CHANGES_REQUESTED** [review]
Didn't quite finish reviewing all the code but thought I should send over my review comments.

File: server/backend/api/app/routers/router_forest_client.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+from fastapi import APIRouter, Query
+
+from .. import schemas
+
+LOGGER = logging.getLogger(__name__)
+
+router = APIRouter()
+
+
+@router.get("/search", response_model=List[schemas.FamForestClient])
```
[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/backend/api/app/routers/router_forest_client.py:15]: Unlike other router that has the authentication check done at "router" for individual endpoint (and can be improved), this is done by include this at higher up at "main.py" for the whole "router" control with **FastAPI** "**dependencies**"
```
app.include_router(router_forest_client.router,
                   prefix=apiPrefix + '/forest_clients',
                   dependencies=[Depends(jwt_validation.authorize)],
                   tags=["FAM Forest Clients"])
```

File: server/backend/api/app/integration/forest_client/forest_client.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import logging
+from http import HTTPStatus
+
+import requests
+from api.config import config
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ForestClient():
```
[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/backend/api/app/integration/forest_client/forest_client.py:10]: Thanks Basil. I don't know either from Python world... but seems to me Python does not borrow "service" that much from various tutorials or documentations.... However, I will rename this, I do like the sound of "ForestClientService".

[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/backend/api/app/integration/forest_client/forest_client.py]: Renamed.

File: server/backend/api/app/integration/forest_client/forest_client.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    def find_by_client_number(self, p_client_number: str):
+        """
+        Find Forest Client(s) information based on p_client_number search query field.
+
+        :param p_client_number: Forest Client Number string (8 digits).
+                                Note! Current Forest Client API can only do exact match.
+                                '/api/clients/findByClientNumber/{clientNumber}'
+        :return: Search result (JSON) for a Forest Client information object.
+        """
+        url = f"{self.api_clients_url}/findByClientNumber/{p_client_number}"
```
[the remark this answers is not in the document]
Comment (ianliuwk1019) [reply server/backend/api/app/integration/forest_client/forest_client.py:38]: In this case I think it is not necessary. Our "router endpoint" for this one does have Pydantic validation built-in (variable Typing and length checks) to request object. And, FAM in this case only concerns the 'search' result from Forest Client API (status ok - with result found or not).

File: server/backend/api/config/config.py [hunk 7]
```diff
@@ -143,3 +144,16 @@ def get_allow_origins():
     allow_origins = [get_env_var("ALLOW_ORIGIN")] if is_on_aws() else ["*"]
     LOGGER.info(f"allow_origins -- {allow_origins}")
     return allow_origins
+
+
+def get_forest_client_api_token():
+    api_token = get_env_var("FC_API_TOKEN") if is_on_aws() \
+        else "4r0dn5wC0JTs6N1tdw6NDzpjBRQwEJ6u"  # Test env.
```
Comment (basilv) [inline server/backend/api/config/config.py:151]: Probably bad security - this token is now exposed to the entire internet.

File: server/backend/api/app/integration/forest_client/forest_client.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                                '/api/clients/findByClientNumber/{clientNumber}'
+        :return: Search result as List for a Forest Client information object.
+                 Current Forest Client API returns exact one result or http status
+                 other than 200 with message content. The intent for FAM search is for
+                 wild card search and Forest Client API could be capable of doing that
+                 in next version.
+        """
+        url = f"{self.api_clients_url}/findByClientNumber/{p_client_number}"
+        LOGGER.debug(f"ForestClientService find_by_client_number() - url: {url}")
+        r = self.session.get(url)
```
Comment (basilv) [inline server/backend/api/app/integration/forest_client/forest_client.py:44]: Two error scenarios:
1. HTTP get fails entirely (no service running) - likely a python error like TCP connection timeout
2. The call returns HTTP status code 5xx indicating error in forest client.

**Review by basilv — CHANGES_REQUESTED** [review]

File: server/backend/api/config/config.py (hunk 7 above)
Comment (ianliuwk1019) [reply server/backend/api/config/config.py:151]: Initially I treated this token in TEST environment isn't needed to be 'secure'.
You finally caught me, security officer!!

File: server/backend/api/app/integration/forest_client/forest_client.py (hunk 8 above)
Comment (ianliuwk1019) [reply server/backend/api/app/integration/forest_client/forest_client.py:44]: Thanks Basil! As discussed, will try to see what happen in python and deal with 5xx error.

File: server/backend/api/config/config.py (hunk 7 above)
Comment (ianliuwk1019) [reply server/backend/api/config/config.py]: I removed the default TEST token now, officer... I will only share it with my team privately (if they give me a chocolate, I like chocolate)

File: server/backend/api/app/integration/forest_client/forest_client.py (hunk 8 above)
Comment (ianliuwk1019) [reply server/backend/api/app/integration/forest_client/forest_client.py:44]: Python doc says: ```In the event of a network problem (e.g. DNS failure, refused connection, etc), Requests will raise a [ConnectionError](https://requests.readthedocs.io/en/latest/api/#requests.ConnectionError) exception``` so same as normal behavior, framework will handle that.
However, python best practice is we need to set "timeout" limit as there is no default one from python.

Comment (ianliuwk1019) [reply server/backend/api/app/integration/forest_client/forest_client.py]: Error handling for 400/500 and network error were addressed and discussed yesterday.
A note from handling 404 from ForestClient API as it uses this code to indicate return of "Client Not Found" so we handle this case in try-catch block and return empty for our search purpose.

File: server/backend/api/config/config.py (hunk 7 above)
Comment (ianliuwk1019) [reply server/backend/api/config/config.py]: Token removed.
Encountered using token from GH report secret for the CI run. Regular terraform plan seems to be fine.
Resolved from CI yaml template.

**Review by basilv — APPROVED** [review]

**Pull request merged**